src/python/WMCore/WorkQueue/WorkQueueBackend.py

```diff
@@ -458,28 +464,21 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
         # Fetch workqueue elements in slices, using the CouchDB "limit" and "skip"
         # options for couch views. Conditions to stop this loop are:
         #  a) have a hard stop at 50k+1 (we might have to make this configurable)
-        #  b) stop as soon as an empty slice is returned by Couch (thus all docs have
-        #     already been retrieve)
+        #  b) stop once total_rows is reached
         #  c) or, once "numElems" elements have been accepted
-        numSkip = 0
-        breakOut = False
-        while True:
-            if breakOut:
-                # then we have reached the maximum number of elements to be accepted
-                break
-            self.logger.info("  with limit docs: %s, and skip first %s docs", docsSliceSize, numSkip)
-            options['skip'] = numSkip
+        numSlices = ceil(numAvail / docsSliceSize)
```

Comment:
Even though the code seems reasonable it is very hard to follow especially since someone needs to track the logic of slices. I rather prefer to get some code re-factoring (since changes are required anyway) and make it clear, e.g.
```
for sliceNum in range(numSlices):
      docs = self.getSliceOfDocs(sliceNum)
      wflows = self.getAvailableWorkflows(docs)
      siteJobCounts = self.getSiteJobCounts(wflows)
return elements, siteJobCounts
```
Such logic will be much easier to read and follow, where individual functions will perform relevant parts of current logic.

Reply:
Yes, I've been considering to separate that block dealing with the elements and checking against siteJobCounts (basically verifying the priority of the work and having the final evaluation of whether the WQE can go through or not).
I will see what can be done tomorrow. But I don't plan to make any major refactorings now, especially because this will have to go through a complete change once we go to CouchDB 4.

Reply:
Okay, the post-processing logic has been separated in its own method.